fixed problem with equal axes

13bzhang opened the pull request:
A solution to Issue #218 "Equal scaling of axes not working" -- so that the height and width ratios will be those specified by users. Conformed to the style guide.

cpsievert commented:
@13bzhang looking good!

I forgot to mention that when we add a new feature, we should add a test to demonstrate that it's working. In this case, I think starting a `test-ggplot-coord.R` file makes sense.

cpsievert commented:
> The message below was automatically generated after build https://travis-ci.org/ropensci/plotly/builds/62757494 

 On TravisCI, commit f6f21054275131fdefb2ff43795192f3420458e3 was successfully merged with 789e1ec11eb939ab9bd054d5384548336aaa0d07 (master) to create 52920bf2f3a38cee8378015a31bb61acf9d697f8. A visual testing table comparing 789e1ec11eb939ab9bd054d5384548336aaa0d07 with 52920bf2f3a38cee8378015a31bb61acf9d697f8 can be found here:
 http://ropensci.github.io/plotly-test-table/tables/52920bf2f3a38cee8378015a31bb61acf9d697f8/index.html

13bzhang commented:
Cool, I will add a test.

13bzhang commented:
I added a testthat [test](https://github.com/ropensci/plotly/blob/baobao-equal_axes/tests/testthat/test-ggplot-coord.R) as Carson suggested.

@cpsievert @mkcor @chriddyp

cpsievert commented:
Looks like the [new tests aren't passing](https://travis-ci.org/ropensci/plotly/builds/62785338#L4152-L4166). There should be a little icon next to each commit to signify whether or not the build was successful. I'm having the same problem over at #217. Not sure why that is.

13bzhang commented:
Hi Carson, Sorry I'm stuck going through commencement weekend. I will take a look at this today.

mkcor commented on tests/testthat/test-ggplot-coord.R:
```diff
@@ -0,0 +1,52 @@
+  y_range <- range(built[[2]]$ranges[[1]]$y.major_source, na.rm = TRUE)
+  yx_ratio <- (y_range[2] - y_range[1]) / (x_range[2] - x_range[1])
+  expect_identical(la$height/la$width, yx_ratio * p$coordinates$ratio)
+})
```
Please add a new line.

13bzhang replied:
Do you mean to add a new line before line 40? I'm confused :confused:

mkcor commented on tests/testthat/test-ggplot-coord.R:
```diff
@@ -0,0 +1,52 @@
+
+# Data where x ranges from 0-10, y ranges from 0-30
+set.seed(202)
+dat <- data.frame(xval = runif(40,0,10), yval = runif(40,0,30))
```
I remember that we decided not to use randomized data in our test suite, so that the image diff's remain meaningful.  But I may have totally missed a discussion on the topic in the meantime...  If so, please let me know.

chriddyp replied:
yeah, that is still correct. I recommend just copying and pasting the results of the random data into the test :)

mkcor replied:
Good. An example of this practice can be found in this test: https://github.com/ropensci/plotly/blob/master/tests/testthat/test-plotly-filename.R

cpsievert replied:
I'm pretty sure `set.seed()` ensures we have the same set of numbers, so image diffs should be meaningful. That being said, I prefer setting the seed over copy/pasting numbers (I think it makes the tests easier to read).

Even better, use a built-in dataset such as `mtcars`, `cars`, `iris`, `diamonds`, etc.

13bzhang replied:
Yes. ```set.seed()``` ensures we get the same data each time. I second
Carson's point. I used the set.seed with the same seed as used in the R
Cookbook. But I can use one of the datasets that come with R if it's
preferred.

Gotta go to graduation now! Will be back shortly.

-- Baobao Zhang

In tests/testthat/test-ggplot-coord.R
<https://github.com/ropensci/plotly/pull/223#discussion_r30524645>:

> +  stopifnot(is.ggplot(gg))
> +  stopifnot(is.numeric(n.traces))
> +  save_outputs(gg, paste0("coord_fixed-", name))
> +  L <- gg2list(gg)
> +  all_traces <- L$data
> +  no_data <- sapply(all_traces, function(tr) {
> +    is.null(tr[["x"]]) && is.null(tr[["y"]])
> +  })
> +  has_data <- all_traces[!no_data]
> +  expect_equal(length(has_data), n_traces)
> +  list(traces = has_data, layout = L$layout)
> +}
> +
> +# Data where x ranges from 0-10, y ranges from 0-30
> +set.seed(202)
> +dat <- data.frame(xval = runif(40,0,10), yval = runif(40,0,30))

I'm pretty sure set.seed() ensures we have the same set of numbers, so
image diffs should be meaningful. That being said, I prefer setting the
seed over copy/pasting numbers (I think it makes the tests easier to read).

Even better, use a built-in dataset such as mtcars, cars, iris, diamonds,
etc.

—
Reply to this email directly or view it on GitHub
<https://github.com/ropensci/plotly/pull/223/files#r30524645>.

mkcor replied:
Thanks, @cpsievert and @13bzhang ! I have learned something today. :)
I agree, `set.seed(1); runif(100)` is definitely easier to read (and diff) than a sequence of 100 numbers...

13bzhang replied:
@mkcor : Does that mean we can keep `set.seed(22)` as it was written in the codebook?

mkcor replied:
Yes, all good!

13bzhang commented:
@cpsievert 
Hey Carson, I fixed the problem with my `expect_traces` function. I think that's what's failing the test. Oops my bad.

cpsievert commented:
> The message below was automatically generated after build https://travis-ci.org/ropensci/plotly/builds/63216395 

 On TravisCI, commit 31a1ef6ce070406c0a854c3a5fbf5a5f93f36ce6 was successfully merged with 789e1ec11eb939ab9bd054d5384548336aaa0d07 (master) to create 147d8a436687c219ba65c91630fbfafe14fe0942. A visual testing table comparing 789e1ec11eb939ab9bd054d5384548336aaa0d07 with 147d8a436687c219ba65c91630fbfafe14fe0942 can be found here:
 http://ropensci.github.io/plotly-test-table/tables/147d8a436687c219ba65c91630fbfafe14fe0942/index.html

mkcor commented on tests/testthat/test-ggplot-coord.R:
```diff
@@ -0,0 +1,52 @@
+  y_range <- range(built[[2]]$ranges[[1]]$y.major_source, na.rm = TRUE)
+  yx_ratio <- (y_range[2] - y_range[1]) / (x_range[2] - x_range[1])
+  expect_identical(la$height/la$width, yx_ratio * p$coordinates$ratio)
+})
```
@13bzhang Sorry, I must have clicked the wrong line: I meant "Please add new line at end of file" so we don't get this red sign from GitHub. Note that if you edit with vim, a new line at end of file is automatically added; in RStudio's editor, you need to actually add a line break after your last line.

13bzhang replied:
@mkcor 
Got it! I added the space to the end of the test.

cpsievert commented:
> The message below was automatically generated after build https://travis-ci.org/ropensci/plotly/builds/63243339 

 On TravisCI, commit ba738b95a9c058fd23fd5db3b1e1cbe51ec6237b was successfully merged with 789e1ec11eb939ab9bd054d5384548336aaa0d07 (master) to create 62801e5b243fdf48d6cc7ec83cea1ac0571dc794. A visual testing table comparing 789e1ec11eb939ab9bd054d5384548336aaa0d07 with 62801e5b243fdf48d6cc7ec83cea1ac0571dc794 can be found here:
 http://ropensci.github.io/plotly-test-table/tables/62801e5b243fdf48d6cc7ec83cea1ac0571dc794/index.html

13bzhang commented:
@cpsievert @mkcor @chriddyp 
Is it ok to merge now?

cpsievert commented:
> The message below was automatically generated after build https://travis-ci.org/ropensci/plotly/builds/63413630 

 On TravisCI, commit 23eb47bb92f959eb3b836fd8f83e1f545332b6cf was successfully merged with 789e1ec11eb939ab9bd054d5384548336aaa0d07 (master) to create ff6f23feead39a52bfebe6f679510526792217c4. A visual testing table comparing 789e1ec11eb939ab9bd054d5384548336aaa0d07 with ff6f23feead39a52bfebe6f679510526792217c4 can be found here:
 http://ropensci.github.io/plotly-test-table/tables/ff6f23feead39a52bfebe6f679510526792217c4/index.html

cpsievert commented on tests/testthat/test-ggplot-coord.R:
```diff
@@ -0,0 +1,53 @@
+  # height-width ratio check
+  built <- ggplot_build2(p)
+  x_range <- range(built[[2]]$ranges[[1]]$x.major_source, na.rm = TRUE)
+  y_range <- range(built[[2]]$ranges[[1]]$y.major_source, na.rm = TRUE)
```
Maybe we should get the ranges directly from the data instead of via `ggplot_build()`?

13bzhang replied:
Wouldn't the range of the data itself be different from the `ggplot_build()` range if people decide to set their own `xlim` and `ylim`? I wanted to account for potential issues with that.

cpsievert replied:
That's true, but you aren't setting `xlim()` or `ylim()` in the test itself, so you currently aren't testing for that...

In general, using the *raw* data as ground truth in tests is a good idea. Suppose ggplot2 internals change at some point and the x range is no longer stored in this location: `built[[2]]$ranges[[1]]$x.major_source`. This test wouldn't necessarily detect any problems since you're basically testing that we map the ggplot object in a certain way (which may be wrong).

13bzhang replied:
Hi Carson,
Good point. Ok, the changes should be made in `ggplotly.R` too righto?
@chriddyp @mkcor @chriddyp

cpsievert replied:
IMO tests shouldn't copy the implementation. Instead, compare inputs to outputs (you shouldn't need to use `ggplot_build()` or anything inside `gg2list()`)

13bzhang commented:
Hi Carson, I changed the range to get it from the data instead of from `ggplot_build2`. See:
https://github.com/ropensci/plotly/blob/79257be14d48b2108fba4e09b2d8ce550740c404/R/ggplotly.R#L775-L776
Also I changed the test to reflect these changes.

cpsievert commented:
No.  IMO tests shouldn't copy the implementation. Instead, compare inputs
to outputs (you shouldn't need to use ggplot_build() or anything inside
gg2list ())
On May 22, 2015 3:56 PM, "Baobao Zhang" <notifications@github.com> wrote:

> Hi Carson, I changed the range to get it from the data instead of from
> ggplot_build2. See:
>
> https://github.com/ropensci/plotly/blob/79257be14d48b2108fba4e09b2d8ce550740c404/R/ggplotly.R#L775-L776
> Also I changed the test to reflect these changes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ropensci/plotly/pull/223#issuecomment-104765312>.
>

cpsievert commented:
Ack, sorry, I made that comment before seeing your most recent changes.
The tests look good!
On May 22, 2015 4:26 PM, "Carson Sievert" <cpsievert1@gmail.com> wrote:

> No.  IMO tests shouldn't copy the implementation. Instead, compare inputs
> to outputs (you shouldn't need to use ggplot_build() or anything inside
> gg2list ())
> On May 22, 2015 3:56 PM, "Baobao Zhang" <notifications@github.com> wrote:
>
>> Hi Carson, I changed the range to get it from the data instead of from
>> ggplot_build2. See:
>>
>> https://github.com/ropensci/plotly/blob/79257be14d48b2108fba4e09b2d8ce550740c404/R/ggplotly.R#L775-L776
>> Also I changed the test to reflect these changes.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/ropensci/plotly/pull/223#issuecomment-104765312>.
>>
>

13bzhang commented:
@cpsievert 
Hi Carson, sorry for the late reply as I am traveling in California to work on a project with some folks at UC Berkeley. Yes, I understand what you're saying now. So I made `ggplotly.R` back to the way it was. For the test, I did the [following](https://github.com/ropensci/plotly/blob/baobao-equal_axes/tests/testthat/test-ggplot-coord.R#L25-L35):

```r
test_that("coord_fixed() is translated to the right height-width ratio", {
  info <- expect_traces(p, 1, "force_equal_scaling")
  tr <- info$traces[[1]]
  la <- info$layout
  expect_identical(tr$type, "scatter")
  # height-width ratio check
  x_range <- range(p$data$xval, na.rm = TRUE)
  y_range <- range(p$data$yval, na.rm = TRUE)
  yx_ratio <- (y_range[2] - y_range[1]) / (x_range[2] - x_range[1])
  expect_identical(la$height/la$width, round(yx_ratio) * p$coordinates$ratio)
})
```
The range ratio from the raw data is going to be a bit off from the `built[[2]]$ranges[[1]]$x.major_source` but the rounded ratio should be the same for each of the two plots in my test. Do you want me to implement an even more general test?
@mkcor @chriddyp

cpsievert commented:
> The message below was automatically generated after build https://travis-ci.org/ropensci/plotly/builds/64571974 

 On TravisCI, commit 5d9deca02f1e9b00752d6c904dd9b05c116c9468 was successfully merged with fb4049180862004ab55f54198e0187ab9a2b5d71 (master) to create a5355cd52f26ad079078aa6b404479c743080dfe. A visual testing table comparing fb4049180862004ab55f54198e0187ab9a2b5d71 with a5355cd52f26ad079078aa6b404479c743080dfe can be found here:
 http://ropensci.github.io/plotly-test-table/tables/a5355cd52f26ad079078aa6b404479c743080dfe/index.html

cpsievert commented:
No worries @13bzhang, thanks for making those changes!

I think `expect_equal()` is a little more natural than `expect_identical()` when comparing quantities. `identical()` respects type (i.e., integer, numeric, etc), which isn't really a concern here, and with `expect_equal()` you can specify the tolerance (see 7cefa02 for an example)

13bzhang commented:
Thanks for the tip. I tried it with `expect_equal` and added tolerance. It worked pretty well. Will commit when I get wifi.

13bzhang commented:
@cpsievert I added the parameter name `tolerance` and made the tolerance a bit smaller. Apparently the Travis CI didn't pass 3 days ago but it just passed now.

cpsievert commented:
> The message below was automatically generated after build https://travis-ci.org/ropensci/plotly/builds/65027665 

 On TravisCI, commit af4fd333010bb56e830310ab720b54b53ce258bc was successfully merged with fb4049180862004ab55f54198e0187ab9a2b5d71 (master) to create 357c7c32de7f1db8d0898059b42568f14ddcfbc1. A visual testing table comparing fb4049180862004ab55f54198e0187ab9a2b5d71 with 357c7c32de7f1db8d0898059b42568f14ddcfbc1 can be found here:
 http://ropensci.github.io/plotly-test-table/tables/357c7c32de7f1db8d0898059b42568f14ddcfbc1/index.html

cpsievert commented:
Ah, shoot, the test table didn't build since there were two builds happening at once. @13bzhang do you see the little "restart build" trigger on travis (for some reason I can't)? If so, please restart the build.

13bzhang commented:
I don't think I have access to the little restart build button because I
don't have enough access or whatever. I did a little hack where I closed
the PR and reopened it, which should trigger a rebuild.

13bzhang commented:
Here she blows! :whale: 
I don't know if that was a really dumb way to restart the build.

cpsievert commented on R/ggplotly.R:
```diff
@@ -829,6 +829,22 @@ gg2list <- function(p) {
+    } else {
+      layout$width <- 600
+      layout$height <- layout$height * (1 / p$coordinates$ratio) * yx_ratio 
+    }
```
Hard-coding the height/width is going to make for unpleasant htmlwidget experience. I think the best way to do it is pass this ratio to the HTMLwidget.resize method and impose the ratio on the height/width there. We should also provide a way to specify this ratio for non-ggplots

cpsievert commented:
fixed axes were added in https://github.com/ropensci/plotly/commit/f9e3f2a361c70d7421b71d9986425ec58f08ccc5
